File: Java/Classes/Organization.java
import java.util.ArrayList;

public class Organization {
    private static Organization instance = null;
    private final String name;
    private final ArrayList<Entity> entityList;
    private final ArrayList<Donator> donatorList;
    private final ArrayList<Beneficiary> beneficiaryList;
    private final RequestDonationList currentDonations;
    private Admin admin;

    // Private constructor for singleton
    private Organization(String name) {
        this.name = name;
        this.entityList = new ArrayList<>();
        this.donatorList = new ArrayList<>();
        this.beneficiaryList = new ArrayList<>();
        this.currentDonations = new RequestDonationList();
    }

    // Singleton getInstance method
    public static Organization getInstance() {
        if (instance == null) {
            instance = new Organization("Relief Organization System");
        }
        return instance;
    }

    public void setAdmin(Admin admin) {
        this.admin = admin;
    }

    public Admin getAdmin() {
        return admin;
    }

    public void addEntity(Entity entity) throws Exceptions.EntityExistsException {
        for (Entity e : entityList) {
            if (e.getId() == entity.getId()) {
                throw new Exceptions.EntityExistsException("Entity with ID " + entity.getId() + " already exists");
            }
            if (e.getName().equals(entity.getName()) && e.getClass().equals(entity.getClass())) {
                throw new Exceptions.EntityExistsException("Entity with name " + entity.getName() + " already exists");
            }
        }
        entityList.add(entity);
    }

    public void removeEntity(Entity entity, User requestingUser) throws Exceptions.InvalidMenuException {
        if (!(requestingUser instanceof Admin)) {
            throw new Exceptions.InvalidMenuException("Only admin can remove entities");
        }
        entityList.remove(entity);
    }

    public Entity getEntity(int id) {
        for (Entity entity : entityList) {
            if (entity.getId() == id) {
                return entity;
            }
        }
        return null;
    }

    public RequestDonation getCurrentDonation(Entity entity) {
        return currentDonations.get(entity.getId());
    }

    public void addCurrentDonation(RequestDonation donation) throws Exceptions.EntityNotFoundException {
        currentDonations.add(donation, this);
    }

    public void insertDonator(Donator donator) throws Exceptions.EntityExistsException {
        for (Donator d : donatorList) {
            if (d.getPhone().equals(donator.getPhone())) {
                throw new Exceptions.EntityExistsException("Donator with phone " + donator.getPhone() + " already exists");
            }
        }
        donatorList.add(donator);
    }

    public void removeDonator(Donator donator) {
        donatorList.remove(donator);
    }

    public void insertBeneficiary(Beneficiary beneficiary) throws Exceptions.EntityExistsException {
        for (Beneficiary b : beneficiaryList) {
            if (b.getPhone().equals(beneficiary.getPhone())) {
                throw new Exceptions.EntityExistsException("Beneficiary with phone " + beneficiary.getPhone() + " already exists");
            }
        }
        beneficiaryList.add(beneficiary);
    }

    public void removeBeneficiary(Beneficiary beneficiary) {
        beneficiaryList.remove(beneficiary);
    }

    public void listEntities() {
        if (entityList.isEmpty()) {
            System.out.println("\nNo entities available");
            return;
        }

        System.out.println("\n=== Available Entities ===");

        System.out.println("\nMaterials:");
        boolean hasMaterials = false;
        for (Entity e : entityList) {
            if (e instanceof Material) {
                hasMaterials = true;
                RequestDonation current = getCurrentDonation(e);
                double quantity = (current != null) ? current.getQuantity() : 0;
                System.out.println("ID: " + e.getId() + " - " + e.getName() + " (Available: " + quantity + ")");
                System.out.println(e.getDetails());
            }
        }
        if (!hasMaterials) {
            System.out.println("No materials available");
        }

        System.out.println("\nServices:");
        boolean hasServices = false;
        for (Entity e : entityList) {
            if (e instanceof Service) {
                hasServices = true;
                RequestDonation current = getCurrentDonation(e);
                double quantity = (current != null) ? current.getQuantity() : 0;
                System.out.println("ID: " + e.getId() + " - " + e.getName() + " (Available hours: " + quantity + ")");
                System.out.println(e.getDetails());
            }
        }
        if (!hasServices) {
            System.out.println("No services available");
        }
    }

    public void listBeneficiaries() {
        if (beneficiaryList.isEmpty()) {
            System.out.println("No beneficiaries registered");
            return;
        }

        System.out.println("\n=== Registered Beneficiaries ===");
        for (Beneficiary b : beneficiaryList) {
            System.out.println("\nBeneficiary: " + b.getName());
            System.out.println("Phone: " + b.getPhone());
            System.out.println("Family Members: " + b.getNoPersons());
            System.out.println("Received Donations:");
            b.getReceivedList().monitor();
        }
    }

    public void listDonators() {
        if (donatorList.isEmpty()) {
            System.out.println("No donators registered");
            return;
        }

        System.out.println("\n=== Registered Donators ===");
        for (Donator d : donatorList) {
            System.out.println("\nDonator: " + d.getName());
            System.out.println("Phone: " + d.getPhone());
            System.out.println("Current Offers:");
            d.getOffersList().monitor();
        }
    }

    public void resetBeneficiaryLists() {
        for (Beneficiary b : beneficiaryList) {
            b.getReceivedList().reset();
        }
        System.out.println("All beneficiary received lists have been reset");
    }

    public String getName() {
        return name;
    }

    public ArrayList<Entity> getEntityList() {
        return new ArrayList<>(entityList);
    }

    public ArrayList<Donator> getDonatorList() {
        return new ArrayList<>(donatorList);
    }

    public ArrayList<Beneficiary> getBeneficiaryList() {
        return new ArrayList<>(beneficiaryList);
    }

    public RequestDonationList getCurrentDonations() {
        return currentDonations;
    }
}

File: C++/cpp_files/Organization.cpp
#include "Organization.h"
#include <iostream>
#include <algorithm>
#include "RequestDonationList.h"
#include "exceptions.h"
#include "Offers.h"
#include "Requests.h"

Organization::Organization(const std::string& name)
    : name(name), admin(nullptr), currentDonations(new RequestDonationList()) {
    currentDonations->setOrganization(this);
}

Organization::~Organization() {
    delete admin;
    delete currentDonations;

    // Clean up entity list
    for (Entity* entity : entityList) {
        delete entity;
    }

    // Clean up donator list
    for (Donator* donator : donatorList) {
        delete donator;
    }

    // Clean up beneficiary list
    for (Beneficiary* beneficiary : beneficiaryList) {
        delete beneficiary;
    }
}

Entity* Organization::findEntity(int entityId) const {
    auto it = std::find_if(entityList.begin(), entityList.end(),
        [entityId](const Entity* e) { return e->getId() == entityId; });
    return it != entityList.end() ? *it : nullptr;
}

Donator* Organization::findDonator(const std::string& phone) const {
    auto it = std::find_if(donatorList.begin(), donatorList.end(),
        [&phone](const Donator* d) { return d->getPhone() == phone; });
    return it != donatorList.end() ? *it : nullptr;
}

Beneficiary* Organization::findBeneficiary(const std::string& phone) const {
    auto it = std::find_if(beneficiaryList.begin(), beneficiaryList.end(),
        [&phone](const Beneficiary* b) { return b->getPhone() == phone; });
    return it != beneficiaryList.end() ? *it : nullptr;
}

Entity* Organization::getEntityById(int entityId) const {
    Entity* entity = findEntity(entityId);
    if (!entity) {
        throw EntityNotFoundException("Entity with ID " + std::to_string(entityId) + " not found");
    }
    return entity;
}

void Organization::setAdmin(Admin* newAdmin) {
    delete admin;
    admin = newAdmin;
}

Admin* Organization::getAdmin() const {
    return admin;
}

void Organization::addEntity(Entity* entity) {
    if (findEntity(entity->getId()) != nullptr) {
        throw EntityExistsException("Entity with ID " + std::to_string(entity->getId()) + " already exists");
    }
    entityList.push_back(entity);
}

void Organization::removeEntity(int entityId) {
    Entity* entity = findEntity(entityId);
    if (entity != nullptr) {
        entityList.erase(std::remove(entityList.begin(), entityList.end(), entity), entityList.end());
        delete entity;
    }
}

void Organization::insertDonator(Donator* donator) {
    if (findDonator(donator->getPhone()) != nullptr) {
        throw UserExistsException("Donator with phone " + donator->getPhone() + " already exists");
    }
    donatorList.push_back(donator);

    if (auto* offersList = dynamic_cast<RequestDonationList*>(donator->getOffersList())) {
        offersList->setOrganization(this);
    }
}

void Organization::removeDonator(const std::string& phone) {
    Donator* donator = findDonator(phone);
    if (donator != nullptr) {
        donatorList.erase(std::remove(donatorList.begin(), donatorList.end(), donator), donatorList.end());
        delete donator;
    }
}

void Organization::insertBeneficiary(Beneficiary* beneficiary) {
    if (findBeneficiary(beneficiary->getPhone()) != nullptr) {
        throw UserExistsException("Beneficiary with phone " + beneficiary->getPhone() + " already exists");
    }
    beneficiaryList.push_back(beneficiary);

    beneficiary->getReceivedList()->setOrganization(this);
    if (auto* requestsList = dynamic_cast<RequestDonationList*>(beneficiary->getRequestsList())) {
        requestsList->setOrganization(this);
    }
}

void Organization::removeBeneficiary(const std::string& phone) {
    Beneficiary* beneficiary = findBeneficiary(phone);
    if (beneficiary != nullptr) {
        beneficiaryList.erase(std::remove(beneficiaryList.begin(), beneficiaryList.end(), beneficiary),
                             beneficiaryList.end());
        delete beneficiary;
    }
}

void Organization::listEntities() const {
    std::cout << "\nMaterials:\n";
    for (const Entity* entity : entityList) {
        if (dynamic_cast<const Material*>(entity) != nullptr) {
            std::cout << entity->toString() << "\n\n";
        }
    }

    std::cout << "\nServices:\n";
    for (const Entity* entity : entityList) {
        if (dynamic_cast<const Service*>(entity) != nullptr) {
            std::cout << entity->toString() << "\n\n";
        }
    }
}

void Organization::listBeneficiaries() const {
    std::cout << "\nBeneficiaries:\n";
    for (const Beneficiary* beneficiary : beneficiaryList) {
        std::cout << "Name: " << beneficiary->getName() << "\n";
        std::cout << "Phone: " << beneficiary->getPhone() << "\n";
        std::cout << "Number of persons: " << beneficiary->getNoPersons() << "\n";
        std::cout << "Received items:\n";
        beneficiary->monitorReceived();
        std::cout << "\n";
    }
}

void Organization::listDonators() const {
    std::cout << "\nDonators:\n";
    for (const Donator* donator : donatorList) {
        std::cout << "Name: " << donator->getName() << "\n";
        std::cout << "Phone: " << donator->getPhone() << "\n\n";
    }
}

RequestDonation* Organization::getRequestDonation(int entityId) const {
    return currentDonations->get(entityId);
}

void Organization::addCurrentDonation(int entityId, double quantity) {
    currentDonations->add(entityId, quantity);
}

void Organization::removeCurrentDonation(int entityId) {
    currentDonations->remove(entityId);
}

void Organization::updateCurrentDonation(int entityId, double quantity) {
    currentDonations->modify(entityId, quantity);
}

void Organization::resetCurrentDonations() {
    currentDonations->reset();
}

void Organization::monitorCurrentDonations() const {
    currentDonations->monitor();
}